File: 0031-next-permutation/0031-next-permutation.java
class Solution {
    public void reverse(int[] nums, int start, int end){
        int n = (start + end);
        for(int i=start; i<n/2; i++){
            int temp = nums[i];
            nums[i] = nums[n - i - 1];
            nums[n - i - 1] = temp; 
        }
        return;
    }

    public void nextPermutation(int[] nums) {
        //1. Find the break point 
        int n = nums.length;
        int breakPoint = -1;

        for(int i=n-2; i>=0; i--){
            if(nums[i] < nums[i+1]){
                breakPoint = i;
                break;
            }
        }
        

        if(breakPoint == -1){
            reverse(nums, 0, n);
            return;
        }

        //Swap the element at break point with the next largest element in th right
        int next = 0;
        for(int i=n-1; i>breakPoint; i--){
            System.out.println(nums[i]);
            if(nums[i] > nums[breakPoint]){
                next = i; 
                break;
            }
        }

        int temp = nums[breakPoint];
        nums[breakPoint] = nums[next];
        nums[next] = temp;

        reverse(nums, breakPoint+1, n);

        return;
    }
}

File: 0031-next-permutation/0031-next-permutation.py
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        pivot = n-2
        
        #Find the first dip
        while pivot >= 0 and nums[pivot] >= nums[pivot + 1]:
            pivot -= 1

        if pivot>=0:
            nxt = len(nums) - 1
            while nums[nxt] <= nums[pivot]:
                nxt-=1
            nums[pivot], nums[nxt] = nums[nxt], nums[pivot]
        
        left = pivot+1
        right = n-1
        
        while left < right:
            nums[left], nums[right] = nums[right], nums[left]
            left+=1
            right-=1

            

